utils/diffing_with_keys.ml

```diff
@@ -0,0 +1,143 @@
+  end)
+module Move = Misc.Stdlib.String.Map
+
+type ('a,'state) partial_edge =
+  | Left of int * 'state * 'a
+  | Right of int * 'state * 'a
+  | Both of 'state * 'a * 'a
+
+let edge key state x y =
+  let kx, ky = key (data x), key (data y) in
+  if kx <= ky then
+    (kx,ky), Left (pos x, state, (x,y))
+  else
+    (ky,kx), Right(pos x,state, (x,y))
+
+let add_edge ex ey = match ex, ey with
+  | ex, None -> Some ex
+  | Left (lpos, lstate, l), Some Right (rpos, rstate,r)
+  | Right (rpos, rstate,r), Some Left (lpos, lstate, l) ->
+      let state = if lpos < rpos then rstate else lstate in
+      Some (Both (state,l,r))
+  | Both _ as b, _ | _, Some (Both _ as b)  -> Some b
+  | l, _ -> Some l
+
+let exchanges ~update ~key state changes =
+  let add (state,(swaps,moves)) d =
+    update d state,
+    match d with
+    | Diffing.Change (x,y,_) ->
+        let k, edge = edge key state x y in
+        Swap.update k (add_edge edge) swaps, moves
+    | Diffing.Insert nx ->
+        let k = key (data nx) in
+        let edge = Right (pos nx, state,nx) in
+        swaps, Move.update k (add_edge edge) moves
+    | Diffing.Delete nx ->
+        let k, edge = key (data nx), Left (pos nx, state, nx) in
+        swaps, Move.update k (add_edge edge) moves
+    | _ -> swaps, moves
+  in
+  List.fold_left add (state,(Swap.empty,Move.empty)) changes
+
+
+let swap key test swaps x y =
+  let kx, ky = key (data x), key (data y) in
+  let key = if kx <= ky then kx, ky else ky, kx in
+  match Swap.find_opt key swaps with
+  | None | Some (Left _ | Right _)-> None
+  | Some Both (state, (ll,lr),(rl,rr)) ->
+      match test state ll rr,  test state lr rl with
+      | Ok _, Ok _ ->
+          Some (mk_pos (pos ll) kx, mk_pos (pos rl) ky)
+      | Error _, _ | _, Error _ -> None
+
+let move key test moves x =
+  let name = key (data x) in
+  match Move.find_opt name moves with
+  | None | Some (Left _ | Right _)-> None
+  | Some Both (state,got,expected) ->
+      match test state got expected with
+      | Ok _ ->
+          Some (Move {name; got=pos got; expected=pos expected})
+      | Error _ -> None
+
+let refine ~key ~update ~test state patch =
+  let _, (swaps, moves) = exchanges ~key ~update state patch in
+  let filter = function
+    | Diffing.Keep _ -> None
+    | Diffing.Insert x ->
+        begin match move key test moves x with
+        | Some _ as move -> move
+        | None -> Some (Insert {pos=pos x;insert=data x})
+        end
+    | Diffing.Delete x ->
+        begin match move key test moves x with
+        | Some _ -> None
+        | None -> Some (Delete {pos=pos x;delete=data x})
+        end
+    | Diffing.Change(x,y, reason) ->
+        match swap key test swaps x y with
+        | Some ((pos1,first),(pos2,last)) ->
+            if pos x = pos1 then
+              Some (Swap { pos = pos1, pos2; first; last})
+            else None
+        | None -> Some (Change reason)
+  in
+  List.filter_map filter patch
```

Comment:
Can you discuss the algorithm here a bit ? 

It seems to me you are implicitly building the graph of position changes where each edge represents a move from position i to position i'. The way you are doing it is a bit ... diluted ? Since you specialized the 2-cycle for swap, and build 2 different maps. If you were doing it explicitly, you would immediately get the refined patch, including arbitrary permutations by finding the cycles.

Reply:
The current algorithm is essentially detecting the 2-cycles of the graph of name changes in the patch (with pairs of addition and deletion considered as a cycle going through a special 𝜔 node). Since any edge of a 2-cycles fully identifies the 2-cycle it might belong to, the algorithm compute a map of partial cycles and update it . Switching to an explicit graph would clearly the way to compute larger cycles.